src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,263 @@
+/*-------------------------------------------------------------------------
+ *
+ * replication_origin_session_utils.c
+ *   Functions for managing replication origin session.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "distributed/replication_origin_session_utils.h"
+#include "distributed/remote_commands.h"
+#include "distributed/metadata_cache.h"
+#include "utils/builtins.h"
+#include "miscadmin.h"
+
+static bool IsRemoteReplicationOriginSessionSetup(MultiConnection *connection);
+
+static bool ExecuteRemoteCommandAndCheckResult(MultiConnection *connection, char *command,
+											   char *expected);
+
+static inline bool IsLocalReplicationOriginSessionActive(void);
+
+PG_FUNCTION_INFO_V1(replication_origin_session_start_no_publish);
+PG_FUNCTION_INFO_V1(replication_origin_session_end_no_publish);
+PG_FUNCTION_INFO_V1(replication_origin_session_is_no_publish);
+
+/*
+ * This variable is used to remember the replication origin id of the current session
+ * before resetting it to DoNotReplicateId in SetupReplicationOriginLocalSession.
+ */
+static RepOriginId originalOriginId = InvalidRepOriginId;
+
+/*
+ * Setting that controls whether replication origin sessions are enabled.
+ */
+bool UseReplicationOriginsForInternalTransfers = false;
+
+
+/* replication_origin_session_start_no_publish starts a new replication origin session
+ * in the local node. This function is used to avoid publishing the WAL records to the
+ * replication slot by setting replication origin to DoNotReplicateId in WAL records.
+ * It remembers the previous replication origin for the current session which will be
+ * used to reset the replication origin to the previous value when the session ends.
+ */
+Datum
+replication_origin_session_start_no_publish(PG_FUNCTION_ARGS)
+{
+	if (!UseReplicationOriginsForInternalTransfers)
+	{
+		PG_RETURN_VOID();
+	}
+	SetupReplicationOriginLocalSession();
```

Comment:
as discussed in the call, we would not want to use this function directly from the UDF because the memory context would get reset after the UDF call is done

Reply:
I have added a boolean parameter "isPartOfRemoteSessionSetup"  to differentiate between direct calls to "SetupReplicationOriginLocalSession" with false and calls through the UDF to true. The contex reset handler will be called only if the parameter is false